tensorcircuit/results/qem/qem_methods.py

```diff
@@ -0,0 +1,358 @@
+        :param spacing: How many identity spacing gates to apply between dynamical
+              decoupling gates, defaults to -1
+        :type spacing: int, optional
+        """
+        dd_sequence = dd_option.rules.general_rule(
+            slack_length=slack_length,
+            spacing=spacing,
+            gates=gates,
+        )
+        return dd_sequence
+
+    if isinstance(rule, list):
+        gates = []
+        for i in rule:
+            gates.append(getattr(cirq, i))
+        rule = dd_rule
+
+    if ignore_idle_qubit is True:
+        qlist = used_qubits(circuit)
+    else:
+        qlist = list(range(circuit.circuit_param["nqubits"]))
+
+    rule_partial = partial(
+        rule, **rule_args
+    )  # rule_args influence the finall DD sequence
+    c2 = circuit
+    c3 = add_dd(c2, rule_partial)
+    c3 = prune_ddcircuit(c3, qlist)
+    if fulldd is True:
+        while c2.to_openqasm() != c3.to_openqasm():
+            c2 = c3
+            c3 = add_dd(c2, rule_partial)
+            c3 = prune_ddcircuit(c3, qlist)
+
+    exp = []
+    for i in range(num_trials):  # type: ignore
+        exp.append(executor(c3))
+
+    if iscount is True:
+        sumcount = dict(
+            functools.reduce(operator.add, map(collections.Counter, exp))  # type:ignore
+        )  # type:ignore
+        result = {k: sumcount[k] / num_trials for k in sumcount.keys()}
+    else:
+        result = np.mean(exp)  # type:ignore
+
+    if full_output is True:
+        result = [result, c3]  # type:ignore
+
+    # def executortc(c):
+    #     c = Circuit.from_qiskit(c, c.num_qubits)
+    #     c = prune_ddcircuit(c,qlist)
+    #     return executor(c)
+    # circuit = circuit.to_qiskit()
+    # result = ddd.execute_with_ddd(
+    #     circuit=circuit,
+    #     executor=executortc,
+    #     rule=rule,
+    #     rule_args=rule_args,
+    #     num_trials=num_trials,
+    #     full_output=full_output,
+    # )
+
+    return result  # type:ignore
+
+
+def rc_candidates(gate: Gate) -> List[Any]:
+    pauli = [m.tensor for m in gates.pauli_gates]
+    if isinstance(gate, gates.Gate):
+        gate = gate.tensor
+    gatem = backend.reshapem(gate)
+    r = []
+    for combo in product(*[range(4) for _ in range(4)]):
+        i = (
+            np.kron(pauli[combo[0]], pauli[combo[1]])
+            @ gatem
+            @ np.kron(pauli[combo[2]], pauli[combo[3]])
+        )
+        if np.allclose(i, gatem, atol=1e-4):
+            r.append(combo)
+        elif np.allclose(i, -gatem, atol=1e-4):
+            r.append(combo)
+    return r
+
+
+def _apply_gate(c: Any, i: int, j: int) -> Any:
+    if i == 0:
+        c.i(j)
+    elif i == 1:
+        c.x(j)
+    elif i == 2:
+        c.y(j)
+    elif i == 3:
+        c.z(j)
+    return c
+
+
+def rc_circuit(c: Any) -> Any:
+    qir = c.to_qir()
+    cnew = Circuit(c.circuit_param["nqubits"])
+    for d in qir:
+        if len(d["index"]) == 2:
+            rc_list = choice(rc_candidates(d["gate"]))
```

Comment:
we better have some cache mechanism here, otherwise, we compute the possible rc candidate for cnot gate for each cnot, could be very slow